domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java

```diff
@@ -1596,91 +1595,207 @@ public TotaledResults(ResultSet set, Integer total) {
-              } catch (Exception e) {
-                sqlRowStream.close();
-                if (!transactionMode) {
-                  connection.close();
-                }
-                log.error(e.getMessage(), e);
-                replyHandler.handle(Future.failedFuture(e));
-              }
-            }).endHandler(v2 -> {
-              if (!transactionMode) {
-                connection.close();
-              }
-              replyHandler.handle(Future.succeededFuture());
-            });
+    try {
+      QueryHelper queryHelper = buildQueryHelper(table,
+        fieldName, wrapper, returnIdField, facets, distinctOn);
+
+      connection.querySingle(queryHelper.countQuery, countQueryResult -> {
+        if (countQueryResult.failed()) {
+          replyHandler.handle(Future.failedFuture(countQueryResult.cause()));
+          return;
+        }
+        ResultInfo resultInfo = new ResultInfo();
+        resultInfo.setTotalRecords(countQueryResult.result().getInteger(0));
+        doStreamGetQuery(connection, queryHelper.selectQuery, resultInfo,
+          clazz, facets, replyHandler);
+      });
+    } catch (Exception e) {
+      log.error(e.getMessage(), e);
+      replyHandler.handle(Future.failedFuture(e));
+    }
+  }
 
+  private <T> void doStreamGetQuery(SQLConnection connection, String selectQuery,
+    ResultInfo resultInfo, Class<T> clazz, List<FacetField> facets,
+    Handler<AsyncResult<PostgresClientStreamResult<T>>> replyHandler) {
+
+    connection.queryStream(selectQuery, stream -> {
+      if (stream.failed()) {
+        log.error(stream.cause().getMessage(), stream.cause());
+        replyHandler.handle(Future.failedFuture(stream.cause()));
+        return;
+      }
+      PostgresClientStreamResult streamResult = new PostgresClientStreamResult(resultInfo);
+
+      SQLRowStream sqlRowStream = stream.result();
+      Promise promise = Promise.promise();
+      ResultsHelper<T> resultsHelper = new ResultsHelper<>(sqlRowStream, clazz);
+      boolean isAuditFlavored = isAuditFlavored(resultsHelper.clazz);
+      Map<String, Method> externalColumnSetters = getExternalColumnSetters(resultsHelper.columnNames,
+        resultsHelper.clazz, isAuditFlavored);
+      // fetch rows and produce ResultInfo when facets are all read
+      sqlRowStream.resultSetClosedHandler(v -> sqlRowStream.moreResults()).handler(r -> {
+        JsonObject row = convertRowStreamArrayToObject(sqlRowStream, r);
+        try {
+          Object objRow = null;
+          // deserializeRow can not determine if count or user object
+          // in case where user T=Object
+          // skip the initial count result when facets are in use
+          if (resultsHelper.offset == 0 && !facets.isEmpty()) {
+            resultsHelper.facet = true;
           } else {
-            if (!transactionMode) {
-              connection.close();
+            objRow = deserializeRow(resultsHelper, externalColumnSetters, isAuditFlavored, row);
```

Comment:
Can you use `objRow = (T) deserializeRow(` to clarify what type we process?

Reply:
Yep